hack/generator/pkg/conversions/property_assignment_function.go

```diff
@@ -271,58 +269,45 @@ func (fn *PropertyAssignmentFunction) generateAssignments(
 
-	receiverContainer, ok := fn.asPropertyContainer(receiver.Type())
-	if !ok {
-		var typeDescription strings.Builder
-		receiver.Type().WriteDebugDescription(&typeDescription, types)
+	var sourceEndpoints map[string]ReadableConversionEndpoint
+	var destinationEndpoints map[string]WritableConversionEndpoint
 
-		return errors.Errorf(
-			"expected receiver TypeDefinition %q to be either resource or object type, but found %q",
-			receiver.Name().String(),
-			typeDescription.String())
-	}
-
-	otherContainer, ok := fn.asPropertyContainer(fn.otherDefinition.Type())
-	if !ok {
-		var typeDescription strings.Builder
-		fn.otherDefinition.Type().WriteDebugDescription(&typeDescription, types)
-
-		return errors.Errorf(
-			"expected other TypeDefinition %q to be either resource or object type, but found %q",
-			fn.otherDefinition.Name().String(),
-			typeDescription.String())
+	switch fn.direction {
+	case ConvertFrom:
+		sourceEndpoints = fn.createReadableEndpoints(fn.otherDefinition.Type())
+		destinationEndpoints = fn.createWritableEndpoints(receiver.Type())
+	case ConvertTo:
+		sourceEndpoints = fn.createReadableEndpoints(receiver.Type())
+		destinationEndpoints = fn.createWritableEndpoints(fn.otherDefinition.Type())
+	default:
+		panic(fmt.Sprintf("unexpected conversion direction %q", fn.direction))
 	}
 
-	receiverProperties := fn.createPropertyMap(receiverContainer)
-	otherProperties := fn.createPropertyMap(otherContainer)
-
 	// Flag receiver name as used
 	fn.knownLocals.Add(receiver.Name().Name())
 
-	for _, receiverProperty := range receiverProperties {
-		otherProperty, ok := otherProperties[receiverProperty.PropertyName()]
-		//TODO: Handle renames
-		if ok {
-			var conv StoragePropertyConversion
-			var err error
-			switch fn.direction {
-			case ConvertFrom:
-				conv, err = fn.createPropertyConversion(otherProperty, receiverProperty)
-			case ConvertTo:
-				conv, err = fn.createPropertyConversion(receiverProperty, otherProperty)
-			default:
-				panic(fmt.Sprintf("unexpected conversion direction %q", fn.direction))
-			}
+	for destinationName, destinationEndpoint := range destinationEndpoints {
+		sourceEndpoint, ok := sourceEndpoints[destinationName]
 
-			if err != nil {
-				// An error was returned, we abort creating conversions for this object
-				return errors.Wrapf(err, "creating conversion for property %q of %q", receiverProperty.PropertyName(), receiver.Name())
-			} else if conv != nil {
-				// A conversion was created, keep it for later
-				fn.conversions[string(receiverProperty.PropertyName())] = conv
-			}
+		if !ok {
+			//TODO: Handle property renames
+			continue
+		}
+
+		// Generate a conversion from one endpoint to another
+		conv, err := fn.createConversion(sourceEndpoint, destinationEndpoint)
+		if err != nil {
+			// An error was returned, we abort creating conversions for this object
+			return errors.Wrapf(
+				err,
+				"creating conversion to %s by %s",
+				sourceEndpoint,
+				destinationEndpoint)
```

Comment:
Are the arguments here the right way around?

Reply:
They need to be swapped; will do that in my next PR.